### ISSUE TITLE
False positive for `functools.cached_property` on a union of enum members

### ISSUE THREAD

**Issue by AlexWaygood** [microsoft/pyright#5696]
**Describe the bug**
Pyright emits false-positive errors if you try to access a `functools.cached_property` on a variable that could be any one of several enum members.

**Code or Screenshots**

The below code reproduces the issue without relying on typeshed's stubs for `functools.cached_property` (but I encountered it trying to use `functools.cached_property`). The bug does not reproduce if you use `@property` instead of `@functools.cached_property`:

```py
import enum
from typing import Any, Callable, Generic, Self, TypeVar, overload

T = TypeVar("T")

class cached_property(Generic[T]):
    """Simplified version of typeshed's stubs for @functools.cached_property"""
    def __init__(self, func: Callable[[Any], T]) -> None: ...
    @overload
    def __get__(self, instance: None, owner: type[Any] | None = None) -> Self: ...
    @overload
    def __get__(self, instance: object, owner: type[Any] | None = None) -> T: ...


class ParserState(enum.IntEnum):
    DSL_START = 0
    PARAMETER_SECTION_START = 1
    PARAMETER_START = 2
    PARAMETER_LEFT_SQUARE_BEFORE = 3
    PARAMETER_GROUP_BEFORE = 4
    PARAMETER_RIGHT_SQUARE_AFTER = 5
    FUNCTION_DOCSTRING = 6

    @cached_property
    def parsing_parameter(self) -> bool:
        return ParserState.PARAMETER_SECTION_START < self < ParserState.FUNCTION_DOCSTRING

class Parser:
    state: ParserState

    def switch(self) -> None:
        match self.state:
            case ParserState.DSL_START:
                ...
            case ParserState.PARAMETER_SECTION_START:
                ...
            # Type of "parsing_parameter" is unknown
            # Cannot access member "parsing_parameter" for type "Literal[ParserState.PARAMETER_START]"
            # Cannot access member "parsing_parameter" for type "Literal[ParserState.PARAMETER_LEFT_SQUARE_BEFORE]"
            # Cannot access member "parsing_parameter" for type "Literal[ParserState.PARAMETER_GROUP_BEFORE]"
            # Cannot access member "parsing_parameter" for type "Literal[ParserState.PARAMETER_RIGHT_SQUARE_AFTER]"
            # Cannot access member "parsing_parameter" for type "Literal[ParserState.FUNCTION_DOCSTRING]"
            case state if state.parsing_parameter:
                ...
            case ParserState.FUNCTION_DOCSTRING:
                ...
            case _ as state:
                raise AssertionError(f"Unexpected state: {state!r}")
```

**VS Code extension or command-line**
I'm running pyright via Pylance v2023.8.20, which I believe means I'm using pyright 1.1.320

**Comment by erictraut** [comment]
Thanks for the bug report. Here's a more minimal example:

```python
from enum import Enum
from functools import cached_property

class MyEnum(Enum):
    A = 0

    @cached_property
    def prop(self) -> bool:
        ...

MyEnum.A.prop  # Cannot access member "prop"
```

Enums in Python are full of special casing and weird / unexpected behaviors, so type checkers need to include a bunch of special casing. This appears to be a bug in some of that special-casing logic. I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.323, which I just published. It will also be included in a future release of pylance.
